src/ccddata.jl

```diff
@@ -0,0 +1,69 @@
+# abstract data type for CCDData
+"""
+    AbstractCCDData{T}
+
+Supertype for `CCDData` based on `AbstractMatrix` interface.
+"""
+abstract type AbstractCCDData{T} <: AbstractMatrix{T} end
+
+# custom data type to hold ImageHDU
+"""
+    CCDData(data::AbstractMatrix, hdr::FITSHeader)
+
+Struct to store `ImageHDU`, derived from [`AbstractCCDData`](@ref).
+"""
+struct CCDData{T,M<:AbstractMatrix{T}} <: AbstractCCDData{T}
+    data::M
+    hdr::FITSHeader
+end
+
+"""
+    CCDData(hdu::ImageHDU)
+
+Loads `ImageHDU` as `CCDData`.
+"""
+CCDData(hdu::ImageHDU) = CCDData(getdata(hdu), read_header(hdu))
+
+"""
+    CCDData(data::AbstractArray)
+
+Creates `CCDData` from an `Array`.
+"""
+CCDData(data) = CCDData(data, get_default_header(data))
+
+"""
+    CCDData(path::String, hdu)
+
+Loads HDU from `hdu` index in `FITS` file at `path` as `CCDData`.
+"""
+function CCDData(path::String, hdu)
+    fh = FITS(path)
+    ccd = CCDData(fh[hdu])
+    close(fh)
+    return ccd
+end
+CCDData{T}(data, hdr) where {T<:Number} = CCDData(T.(data), hdr)
+
+# extending the AbstractMatrix interface
+Base.size(ccd::CCDData) = size(ccd.data)
```

Comment:
from this line on down, you should probably switch to using `AbstractCCDData`. As I wrote before, if you define the interface like this
```julia
data(ccd::CCDData) = ccd.data
hdr(ccd::CCDData) = ccd.hdr
```
you should have no problems.

E.G

```suggestion
Base.size(ccd::AbstractCCDData) = size(data(ccd))
```


If you don't do this, then `AbstractCCDData` is essentially worthless, as no methods have been defined for it, meaning any downstream users will have to completely implement _the exact same code that is being implemented here_.

Reply:
Sure, on it!

Reply:
I think the functions `copy`, `convert`, `view`, `selectdim` and broadcast mechanics cannot be generalised for `AbstractCCDData`. I would like to know your thoughts about the current implementation, what more modifications can we do?
I tried to generalise broadcast mechanics and that lead to a lot of errors.

Reply:
Yep, those things are probably going to unavoidable. What are the errors?

Reply:
All lines involving concrete types errors:
```
Base.copy(ccd::CCDData) = CCDData(copy(data(ccd)), deepcopy(hdr(ccd)))
Base.convert(::Type{CCDData{T}}, ccd::CCDData{V}) where {T,V} = CCDData{T}(data(ccd), hdr(ccd))
Base.view(ccd::CCDData, inds...) = CCDData(view(data(ccd), inds...), hdr(ccd))
Base.selectdim(ccd::CCDData, d::Integer, idxs) = CCDData(selectdim(data(ccd), d, idxs), hdr(ccd))
```
and also the broadcast mechanics part:
```
function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{CCDData}}, ::Type{T}) where T
    ccd = find_ccd(bc)
    CCDData(similar(data(ccd), T, axes(bc)), hdr(ccd))
end
```
Because we are creating a new `CCDData` inside it, so we cannot substitute `AbstractCCDData` in place of `CCDData`

Reply:
okay, that sounds about right, but give this a try for the broadcasting:
```julia
function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{CCD}}, ::Type{T}) where {T,CCD<:AbstractCCDData}
    ccd = find_ccd(bc)
    CCD(similar(data(ccd), T, axes(bc)), hdr(ccd))
end
```

Reply:
This part works pretty well:
```
Base.BroadcastStyle(::Type{<:CCDData}) = Broadcast.ArrayStyle{CCDData}()
function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{CCD}}, ::Type{T}) where {T, CCD<:AbstractCCDData}
    ccd = find_ccd(bc)
    CCD(similar(data(ccd), T, axes(bc)), hdr(ccd))
end
"`A = find_ccd(As)` returns the first CCDData among the arguments."
find_ccd(bc::Base.Broadcast.Broadcasted) = find_ccd(bc.args)
find_ccd(args::Tuple) = find_ccd(find_ccd(args[1]), Base.tail(args))
find_ccd(x) = x
find_ccd(::Tuple{}) = nothing
find_ccd(a::AbstractCCDData, rest) = a
find_ccd(::Any, rest) = find_ccd(rest)
```

I had trouble generalising this line:
```
Base.BroadcastStyle(::Type{<:CCDData}) = Broadcast.ArrayStyle{CCDData}()
```

I tried this:
```
Base.BroadcastStyle(::Type{T}) where {T<:AbstractCCDData} = Broadcast.ArrayStyle{T}()
```
But that did not work. Any work around for this?

Reply:
For now, let's keep broadcast mechanics specific to subtype, we can discuss about this later!